pose_prediction/plugin.cpp

```diff
@@ -99,9 +156,203 @@ class pose_prediction_impl : public pose_prediction {
 		Eigen::Quaternionf raw_o (pose.orientation.w(), -pose.orientation.y(), pose.orientation.z(), -pose.orientation.x());
 
 		swapped_pose.orientation = apply_offset(raw_o);
+        swapped_pose.sensor_time = pose.sensor_time;
 
         return swapped_pose;
     }
+
+    // Slightly modified copy of OpenVINS method found in propagator.cpp
+    // Returns a pair of the predictor state_plus and the time associated with the
+    // most recent imu reading used to perform this prediction.
+    std::pair<Eigen::Matrix<double,13,1>,time_type> predict_mean_rk4(double dt) const {
+
+        // Pre-compute things
+        const imu_biases_type *imu_biases = _m_imu_biases->get_latest_ro();
+
+        Eigen::Vector3d w_hat =imu_biases->w_hat;
+        Eigen::Vector3d a_hat = imu_biases->a_hat;
+        Eigen::Vector3d w_alpha = (imu_biases->w_hat2-imu_biases->w_hat)/dt;
+        Eigen::Vector3d a_jerk = (imu_biases->a_hat2-imu_biases->a_hat)/dt;
+
+        // y0 ================
+        Eigen::Vector4d q_0 = Eigen::Matrix<double,4,1>{imu_biases->pose(0), imu_biases->pose(1), imu_biases->pose(2), imu_biases->pose(3)};
+        Eigen::Vector3d p_0 = Eigen::Matrix<double,3,1>{imu_biases->pose(4), imu_biases->pose(5), imu_biases->pose(6)};
+        Eigen::Vector3d v_0 = Eigen::Matrix<double,3,1>{imu_biases->pose(7), imu_biases->pose(8), imu_biases->pose(9)};
+
+        // k1 ================
+        Eigen::Vector4d dq_0 = {0,0,0,1};
+        Eigen::Vector4d q0_dot = 0.5*Omega(w_hat)*dq_0;
+        Eigen::Vector3d p0_dot = v_0;
+        Eigen::Matrix3d R_Gto0 = quat_2_Rot(quat_multiply(dq_0,q_0));
+        Eigen::Vector3d v0_dot = R_Gto0.transpose()*a_hat-Eigen::Vector3d{0.0, 0.0, 9.81};
+
+        Eigen::Vector4d k1_q = q0_dot*dt;
+        Eigen::Vector3d k1_p = p0_dot*dt;
+        Eigen::Vector3d k1_v = v0_dot*dt;
+
+        // k2 ================
+        w_hat += 0.5*w_alpha*dt;
+        a_hat += 0.5*a_jerk*dt;
+
+        Eigen::Vector4d dq_1 = quatnorm(dq_0+0.5*k1_q);
+        //Eigen::Vector3d p_1 = p_0+0.5*k1_p;
+        Eigen::Vector3d v_1 = v_0+0.5*k1_v;
+
+        Eigen::Vector4d q1_dot = 0.5*Omega(w_hat)*dq_1;
+        Eigen::Vector3d p1_dot = v_1;
+        Eigen::Matrix3d R_Gto1 = quat_2_Rot(quat_multiply(dq_1,q_0));
+        Eigen::Vector3d v1_dot = R_Gto1.transpose()*a_hat-Eigen::Vector3d{0.0, 0.0, 9.81};
+
+        Eigen::Vector4d k2_q = q1_dot*dt;
+        Eigen::Vector3d k2_p = p1_dot*dt;
+        Eigen::Vector3d k2_v = v1_dot*dt;
+
+        // k3 ================
+        Eigen::Vector4d dq_2 = quatnorm(dq_0+0.5*k2_q);
+        //Eigen::Vector3d p_2 = p_0+0.5*k2_p;
```

Comment:
Nit: remove this extraneous comment

Reply:
This was lifted from OpenVINS, so I'm thinking not modifying will reduce the diff-drift of our copy from the original.